### PR TITLE
drainer: fix ts map

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
### What problem does this PR solve? 
fix issue: https://internal.pingcap.net/jira/browse/TOOL-1296

### What is changed and how it works?
get ts from downstream after batch is executed.

### Check List 

Tests 

 - Unit test
 - Manual test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: pkg/loader/load_test.go [hunk 1]
```diff
@@ -473,3 +475,38 @@ func (s *runSuite) TestShouldFlushWhenInputIsEmpty(c *check.C) {
...
+	updateLastFinishTSInterval = 0
+	defer func() {
+		updateLastFinishTSInterval = originUpdateLastFinishTSInterval
+	}()
+	txns = []*Txn{
+		{Metadata: 7},
+		{Metadata: 8},
+	}
+	loader.markSuccess(txns...)
+	c.Assert(txns[len(txns)-1].FinishTS, check.Equals, int64(88881234))
```
Comment (suzaku) [inline pkg/loader/load_test.go]: The original assertion is to make sure that when we call `markSuccesses` again no `FinishTS` are set, so it should be `0`.

Comment (WangXiangUSTC) [reply pkg/loader/load_test.go]: updated in https://github.com/pingcap/tidb-binlog/pull/670/commits/8b1e5bde976da233ddbcdd4ef36291ef59b421ba

**Comment by suzaku** [comment]
LGTM

File: drainer/sync/mysql.go [hunk 2]
```diff
@@ -38,14 +38,14 @@ type MysqlSyncer struct {
...
-func NewMysqlSyncer(cfg *DBConfig, tableInfoGetter translator.TableInfoGetter, worker int, batchSize int, queryHistogramVec *prometheus.HistogramVec, sqlMode *string) (*MysqlSyncer, error) {
+func NewMysqlSyncer(cfg *DBConfig, tableInfoGetter translator.TableInfoGetter, worker int, batchSize int, queryHistogramVec *prometheus.HistogramVec, sqlMode *string, destDBType string) (*MysqlSyncer, error) {
 	db, err := createDB(cfg.User, cfg.Password, cfg.Host, cfg.Port, sqlMode)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 
 	var opts []loader.Option
-	opts = append(opts, loader.WorkerCount(worker), loader.BatchSize(batchSize))
+	opts = append(opts, loader.WorkerCount(worker), loader.BatchSize(batchSize), loader.DownstreamTp(destDBType))
```
Comment (GregoryIan) [inline drainer/sync/mysql.go]: is definition of `DownstreamTp` tricky?  you adjust type of downstream, and only want to use it to check whether we need to keep slave TS.  I suggest that  it's better not to covert it to other type, it only make some people confused

File: drainer/sync/syncer.go [hunk 3]
```diff
@@ -24,6 +24,9 @@ type Item struct {
 	PrewriteValue *pb.PrewriteValue // only for DML
 	Schema        string
 	Table         string
+
+	// the finish TS executed in downstream, only for tidb
+	FinishTS int64
```
Comment (GregoryIan) [inline drainer/sync/syncer.go]: is `AppliedTS` better?

File: drainer/syncer.go [hunk 4]
```diff
@@ -203,9 +206,10 @@ func (s *Syncer) handleSuccess(fakeBinlog chan *pb.Binlog, lastTS *int64) {
 				atomic.StoreInt64(lastTS, ts)
 			}
 
-			// save ASAP for DDL
-			if item.Binlog.DdlJobId > 0 {
+			// save ASAP for DDL, and if FinishTS > 0, we should save the ts map
+			if item.Binlog.DdlJobId > 0 || item.FinishTS > 0 {
 				saveNow = true
+				finishTS = item.FinishTS
```
Comment (GregoryIan) [inline drainer/syncer.go]: where to reset `finishTS`

File: pkg/loader/load.go [hunk 5]
```diff
@@ -181,6 +206,10 @@ func (s *loaderImpl) GetSafeMode() bool {
 }
 
 func (s *loaderImpl) markSuccess(txns ...*Txn) {
```
Comment (GregoryIan) [inline pkg/loader/load.go]: does `markSuccess` take effect  synchronously or asynchronously?

Comment (WangXiangUSTC) [reply pkg/loader/load.go]: `markSuccess` can take effect asynchronously, but if we get applied ts here, can only run synchronously after execute DML and DDL

File: drainer/syncer.go (hunk 4 above)
Comment (WangXiangUSTC) [reply drainer/syncer.go]: updated https://github.com/pingcap/tidb-binlog/pull/670/commits/4c8782f0e021d79d1b1ecb9094716dc0a9810c2f

File: drainer/sync/syncer.go (hunk 3 above)
Comment (WangXiangUSTC) [reply drainer/sync/syncer.go]: updated https://github.com/pingcap/tidb-binlog/pull/670/commits/4c8782f0e021d79d1b1ecb9094716dc0a9810c2f

File: drainer/sync/mysql.go (hunk 2 above)
Comment (WangXiangUSTC) [reply drainer/sync/mysql.go]: updated https://github.com/pingcap/tidb-binlog/pull/670/commits/412a5226bcd11cac6e2422db7c47e3eecc8a6306

**Comment by july2993** [comment]
use github.com/pingcap/parser/mysql instead of github.com/pingcap/tidb/mysql
@WangXiangUSTC pls fix ci

**Comment by WangXiangUSTC** [comment]
@GregoryIan @july2993 PTAL

**Comment by GregoryIan** [comment]
LGTM
and  @july2993 PTAL

**Pull request merged**